salt/_utils/kubernetes_utils.py

```diff
@@ -0,0 +1,535 @@
+        except (TypeError, ValueError):
+            raise _type_error(value, expected='a date-time string')
+
+    dict_match = DICT_PATTERN.match(type_string)
+    if dict_match is not None:
+        if not isinstance(value, dict):
+            raise _type_error(value, expected='a dictionary')
+
+        value_type_str = dict_match.group('value_type')
+        return {
+            key: _cast_value(val, value_type_str)
+            for key, val in value.items()
+        }
+
+    list_match = LIST_PATTERN.match(type_string)
+    if list_match is not None:
+        if not isinstance(value, list):
+            raise _type_error(value, expected='a list')
+
+        value_type_str = list_match.group('value_type')
+        return [_cast_value(val, value_type_str) for val in value]
+
+    try:
+        model = getattr(k8s_client.models, type_string)
+    except AttributeError:
+        # This should never happen, otherwise this function should get updated
+        raise ValueError(
+            'Unknown type string provided: {}.'.format(type_string)
+        )
+
+    if not isinstance(value, dict):
+        raise _type_error(
+            value,
+            expected='a dict to cast as a "{}"'.format(model.__name__)
+        )
+
+    return _build_standard_object(model, value)
+
+
+def _type_error(value, expected):
+    return TypeError(
+        'Expected {}, received: {} (type: {})'.format(
+            expected, value, type(value)
+        )
+    )
+
+
+def validate_manifest(manifest):
+    """Ensure a Kubernetes object is strictly conformant to its OpenAPI schema.
+
+    This method relies on the K8s client models, which are generated from said
+    schema. The approach is to cast the original `manifest` dict into such
+    models, and then obtain the dict representation from the model back again,
+    ensuring that nothing is lost from the original manifest (the conversion
+    also checks that no "unsupported" field is provided).
+
+    For CustomResources, this only checks that the apiVersion and kind are
+    registered in this file.
+    """
+    obj = convert_manifest_into_object(manifest)
+    result = obj.to_dict()
+
+    # We change camelCase into snake_case for all keys in the source
+    # manifest, as is done by the `kubernetes` client.
+    desired = _cast_dict_keys(manifest, key_cast=_camel_to_snake)
+
+    # We only check fields provided in the source, since the conversion
+    # may add empty or default values for optional fields
+    dictdiff = recursive_diff(desired, result)
+    return not dictdiff.removed() and not dictdiff.changed()
+
+
+def _cast_dict_keys(data, key_cast):
+    if isinstance(data, dict):
+        return {
+            key_cast(key): _cast_dict_keys(value, key_cast)
+            for key, value in data.items()
+        }
+
+    if isinstance(data, list):
+        return [_cast_dict_keys(value, key_cast) for value in data]
+
+    return data
+
+
+def _camel_to_snake(key):
```

Comment:
Not entirely sure it's the exact same, but:

```
'_'.join(p.lower() for p in re.split(r'([A-Z][a-z0-9]*)', 'camelCaseValue') if p)
```

Reply:
Will try the regex approach, and will include doctests!

Reply:
OK, so after going through all attributes in the models from `kubernetes.client.models`, I found that the rules are quite strange.

Here's what I ended up with:

```python
def _convert_attribute_name(key):
    if keyword.iskeyword(key):
        return '_{}'.format(key)

    if key.startswith('$'):
        # Only two supported values, '$ref' and '$schema'
        return key[1:]

    for pattern in ['([a-z])([A-Z0-9])', '([A-Z0-9]+)([A-Z0-9][a-z])']:
        key = re.sub(pattern, r'\1_\2', key)

    return key.lower()
```

To check that it matches all examples across the original models, here's what I did:
```python
attr_map = {}
for model in kubernetes.client.models.__dict__:
    attr_map.update(getattr(model, 'attribute_map', {}))

assert all(
    python_key == _convert_attribute_name(k8s_key)
    for python_key, k8s_key in attr_map.items()
)
```